### DISCUSSION TITLE
Update per API starten / Welche Geräte sind bereit für ein Update?

### DISCUSSION THREAD

**Discussion by SukramJ** [OpenCCU/OpenCCU#2314] [🇩🇪 Fragen&Antworten] [answered]
Per JSON API (`Interface.updateFirmware`)kann ich eine Update auf einem Gerät starten, so zumindest meine Vermutung.

ABER:
Wie kann ich per API herausfinden welches Gerät für ein Update bereit ist?

**Comment by jens-maus** [comment, marked as answer]
Das ist leider von HomeMatic vs. homematicIP unterschiedlich. Siehe hier:

https://github.com/eq-3/occu/blob/master/WebUI/www/config/ic_deviceFirmwareOverview.cgi#L133-L181

Bei BidCos/HomeMatic gibt es `AVAILABLE_FIRMWARE` vs. `FIRMWARE` in den device descriptions und man muss wohl vergleichen und auch `UPDATEABLE` wohl auswerten. Und bei homematicIP gibt es wohl `FIRMWARE_UPDATE_STATE` und dieses kann mehrere Werte von `PERFORMING_UPDATE` bis hin zu `LIVE_DELIVER_FIRMWARE_IMAGE` annehmen.

Und via JSON API sollte man diese Infos via `Interface.getDeviceDescription` erhalten. Siehe:

https://github.com/eq-3/occu/blob/master/WebUI/www/api/methods/interface/dispatch.tcl#L78-L93
https://github.com/eq-3/occu/blob/master/WebUI/www/api/methods/interface/getdevicedescription.tcl

Hoffe das hilft dir weiter!

**Reply by SukramJ** [reply]
Vielen Dank für die schnelle und ausführliche Antwort.
Ich hatte ehrlich gesagt gehofft, dass ich die Infos mit nur einem API Call abrufen kann, und ich nur die Methode übersehen habe.
Da wir die device descriptions in `hahomematic` aus Performancegründen gecacht haben, kriegen wir da auch keine Aktualisierung beim Starten, so das der Updatecheck alle device descriptions laden und prüfen müsste.

Vielleicht kann jemand mit Script-Expertise (@Baxxy13 ??) mit einem Rega-Script unterstützen?

**Reply by SukramJ** [reply]
Ich werde erst mal mit der XmlRPC Methode listDevices arbeiten. Mal schauen wie weit ich damit komme.

**Reply by jens-maus** [reply]
Alles klar. Aber bitte immer den DutyCycle im Auge behalten. Wir hatten in der Vergangenheit ja bei gewissen XMLRPC Abfragen ein sprunghaften anstieg des Funkverkehres...

**Reply by SukramJ** [reply]
Da die Json API hier nur als Wrapper dient sehe ich da keinen Unterschied. Gibt es einen Gerätezugriff beim Zugriff auf die device descriptions?

**Reply by jp112sdl** [reply]
> Vielleicht kann jemand mit Script-Expertise (@Baxxy13 ??) mit einem Rega-Script unterstützen?

Bin zwar nicht Baxxy... aber vielleicht hilft dir dieser Thread weiter: https://homematic-forum.de/forum/viewtopic.php?t=77314&p=750234#p750234

**Reply by SukramJ** [reply]
Sieht interessant aus. Danke.

**Comment by Baxxy13** [comment]
Wie ist denn die Umsetzung seitens Integration angedacht?
Ein Service den der User zykl. starten kann?

Sinn macht so eine Abfrage ja nur für IP-Funk Geräte. Alle anderen sind ja quasi Instant updatebereit. Und wenn der User schon auf der WebUI ist um das Update-File hochzuladen dann kann er auch gleich den Button drücken. 

Für IP würde ich dann über alle Devices am HmIP-RF Interface iterieren. Also etwa so wie hinter dem Link zum Forum.

**Reply by SukramJ** [reply]
HA bietet seit einiger Zeit `Update` Entitäten in denen das Ganze abgebildet werden kann, inkl. dem Update durchführen.
Ein zyklisch laufender Service würde die Daten bei der CCU abfragen und die Entitäten befüllen.
Je nach dem wie der Service die Daten einsammeln kann, ob es Auswirkungen auf den Dutycycle gibt, bzw. abhängig von der Schnittstelle, muss die Updatefrequenz eingestellt werden.

**Reply by SukramJ** [reply]
Der aktuelle Stand kann ab Beta 1.37.0b1 getestet werden.
Das grundsätzlich Vorgehen hab ich in der [Readme](https://github.com/danielperna84/custom_homematic#updating-a-device-firmware) beschrieben.

Testen konnte ich allerdings nur mit HMIP Geräten.
